### PR TITLE
Begin LoadModule implementation (ref #1)

### PR DESCRIPTION
**Pull request by phated**
@ry as asked, here's a PR with the implementation (cleaned up) and tests.   I commented on the missing dependency test about the SIGABRT currently.   I've also allowed "edits from maintainers" if you need to tweak anything.

### CODE REVIEW

File: binding.cc [hunk 1]
```diff
@@ -151,6 +154,78 @@ int worker_load(worker* w, char* name_s, char* source_s) {
   return 0;
 }
 
+
+MaybeLocal<Module> ResolveCallback(Local<Context> context, Local<String> name, Local<Module> referrer) {
```
Comment (ry) [inline binding.cc]: I'm getting a warning here 
```
binding.cc:158:20: warning: 'ResolveCallback' has C-linkage specified, but returns incomplete type 'MaybeLocal<v8::Module>' which could be incompatible with C [-Wreturn-type-c-linkage]
```
It is fixed if this function is moved outside the `extern "C"`

File: worker_test.go [hunk 2]
```diff
@@ -217,6 +217,33 @@ func TestRequestFromJS(t *testing.T) {
...
+func TestModules(t *testing.T) {
+	worker := New(func(msg []byte) []byte {
+		t.Fatal("shouldn't recieve Message")
+		return nil
+	})
+	err1 := worker.LoadModule("dependency.js", `export const test = "ready";`)
+	if err1 != nil {
+		t.Fatal(err1)
+	}
+	err2 := worker.LoadModule("code.js", `import { test } from "dependency.js"; V8Worker2.print(test);`)
```
Comment (ry) [inline worker_test.go]: Could you wrap these long lines? I prefer the code to be at 80cols

File: worker_test.go [hunk 3]
```diff
@@ -217,6 +217,33 @@ func TestRequestFromJS(t *testing.T) {
...
+	}
+}
+
+func TestModulesMissingDependency(t *testing.T) {
+	// I believe this SIGABRTs currently
+	worker := New(func(msg []byte) []byte {
+		t.Fatal("shouldn't recieve Message")
+		return nil
+	})
+	err := worker.LoadModule("code.js", `import { test } from "missing.js"; V8Worker2.print(test);`)
```
Comment (ry) [inline worker_test.go]: long line

File: worker_test.go [hunk 4]
```diff
@@ -217,6 +217,33 @@ func TestRequestFromJS(t *testing.T) {
...
+		t.Fatal(err1)
+	}
+	err2 := worker.LoadModule("code.js", `import { test } from "dependency.js"; V8Worker2.print(test);`)
+	if err2 != nil {
+		t.Fatal(err2)
+	}
+}
+
+func TestModulesMissingDependency(t *testing.T) {
+	// I believe this SIGABRTs currently
```
Comment (ry) [inline worker_test.go]: I will check into it soon.

File: binding.cc (hunk 1 above)
Comment (phated) [reply binding.cc]: Awesome. Will do.

**Comment by phated** [comment]
> Do you also experience that?

Yeah, that's the issue I've been seeing.  I think this is due to storing a pointer in the set instead of the actual `Persistent` - maybe V8 is cleaning it up somehow?

**Comment by phated** [comment]
I'm going to try working on this some more today.  Hopefully I get a better grasp on where the modules are being lost.

**Comment by ry** [comment]
@phated Sounds good. I think with this project ample testing is going to be necessary.

Also check out my other project deno which uses v8worker2. There I can support ES modules because I use the TS compiler. But currently the output of the compiler is AMD format:
https://github.com/ry/deno/blob/c0cc240810f9280ca458c54b4cb69acc30e47e27/runtime.ts#L185
It would be great to instead to output ES2015 modules ([which TS supports](https://github.com/Microsoft/TypeScript/blob/2226cd6a992b699e05de67a99f27669fe71cdf7d/lib/typescript.d.ts#L2477-L2478)) which might enable some performance optimizations.

**Comment by phated** [comment]
@ry soooooo, I got it working 😛

**Comment by phated** [comment]
And now I've fixed the module failure stuff too. (There's a TODO in there that you can probably answer)

**Comment by phated** [comment]
@ry I think this might have gotten lost in all the spam on deno last night so I just wanted to ping it again.

**Comment by phated** [comment]
@ry Heads up, I'm experimenting with `LoadModule` in https://github.com/phated/reasonable (mostly because I really dislike TypeScript) - Current implementation seems a little weird because I need to do an AST parse on the JS code and enumerate the dependencies before loading said module.  It seems like a much different strategy than Deno because you are able to define the AMD factory functions.

**Comment by ry** [comment]
@phated rad!

File: binding.cc [hunk 5]
```diff
@@ -90,6 +92,28 @@ void ExitOnPromiseRejectCallback(PromiseRejectMessage promise_reject_message) {
   exit(1);
 }
 
+MaybeLocal<Module> ResolveCallback(Local<Context> context,
+                                   Local<String> name,
```
Comment (ry) [inline binding.cc]: Nit: use `specifier` here instead of `name`.

File: binding.cc [hunk 6]
```diff
@@ -90,6 +92,28 @@ void ExitOnPromiseRejectCallback(PromiseRejectMessage promise_reject_message) {
...
+                                   Local<Module> referrer) {
+  auto isolate = Isolate::GetCurrent();
+  worker* w = (worker*)isolate->GetData(0);
```
Comment (ry) [inline binding.cc]: I think you should add 
`HandleScope handle_scope(isolate);`
here?

File: worker_test.go [hunk 7]
```diff
@@ -217,6 +217,41 @@ func TestRequestFromJS(t *testing.T) {
 	}
 }
 
+func TestModules(t *testing.T) {
+	worker := New(func(msg []byte) []byte {
+		t.Fatal("shouldn't recieve Message")
+		return nil
+	})
+	exportCode := `
```
Comment (ry) [inline worker_test.go]: s/exportCode/dependencyCode
Or better yet, don't name the string and just put it into the function call below

File: worker_test.go [hunk 8]
```diff
@@ -217,6 +217,41 @@ func TestRequestFromJS(t *testing.T) {
...
+		export const test = "ready";
+	`
+	err1 := worker.LoadModule("dependency.js", exportCode)
+	if err1 != nil {
+		t.Fatal(err1)
+	}
+	importCode := `
```
Comment (ry) [inline worker_test.go]: > don't name the string and just put it into the function call below

same here

**Comment by phated** [comment]
Don't the modules need to be evaluated to work as dependencies?  I believe I ran into that while I was hacking on stuff.  I think we actually need to have a "resolver" function passed to the LoadModule API (or maybe attached to the isolate instance) that gets called before `ResolveCallback` looks up the module in the cache.

File: binding.cc (hunk 6 above)
Comment (phated) [reply binding.cc]: I don't actually understand what handle_scope does. Haha.

File: worker_test.go (hunk 7 above)
Comment (phated) [reply worker_test.go]: I believe this was attempting to shorten the lines.  I can inline it.

**Comment by phated** [comment]
@ry I've updated based on the feedback.  I'm also trying to add a "ModuleResolveCallback" but I'm not sure how to expose a golang func to the C code - I followed the pattern of `recvCb` and looked into cgo but my func still doesn't seem available.  Maybe I need to rebuild some headers?

**Comment by ry** [comment]
closing in favor of #8

**Pull request closed without merging**